### DISCUSSION TITLE
Few queries regarding generic and  Dictionary<string, object> support

### DISCUSSION THREAD

**Discussion by jrkumar-conga** [microsoft/RulesEngine#244] [Q&A]
Hi @abbasc52 
**Q1.Whether we can call any method with generic return type inside a expression?**
for e.g if a method is like  `public T GetFieldValue<T>(string fieldName)` 
Is it possible to write in expression like `"input1.GetFieldValue<string>(\"somefield\")" ? `
because when we try this getting exception `System.Linq.Dynamic.Core.Exceptions.ParseException: 'No property or field 'GetCustomFieldValue' exists in type 'MyObject'`

Also
**Q2. Whether Dictionary<string, object> can be sent as input?**
I know Dictionary<string, object> is supported, however facing the problem to convert each value to it's type before using it in expression.
For e.g. 
```
var inputDict = new Dictionary<string, object> {
                    { "Name", "SomeName" },
                    { "Id", Guid.NewGuid() },
                    { "StartDate", DateTime.Now },
                    { "EndDate", DateTime.Now.AddMonths(2) },
                    { "Price", 1000 },
                };
```
Now if the expression is like `"input1.Price * 0.2"` getting below exception
`System.Linq.Dynamic.Core.Exceptions.ParseException: 'Operator '*' incompatible with operand types 'Object' and 'Double''`
 Same thing with DateTime also.

Our expressions are stored in DB so it will be difficult to parse and change each expression on runtime.

Please let me know if you want more information.

**Comment by abbasc52** [comment]
Hi @jrkumar-conga 

> Q1.Whether we can call any method with generic return type inside an expression?

Currently support for generics is limited. So, a generic return method is not possible currently


> **Q2. Whether Dictionary<string, object> can be sent as input?**

You can send Dictionary<string, object> as input but it can't assume type on it own.
If you need to use certain type, you can cast it as follows: - `TypeName(input)`
e.g. `int(input1.Price) * 0.2`
if you don't want to do it in expression, try mapping the dictionary<string,object> to ExpandoObject before passing as input

```c#
var inputData = new Dictionary<string,object>{
    {"cost",1}
};

dynamic dynamicInputData = new ExpandoObject();
foreach(var kv in inputData){
    (dynamicInputData as IDictionary<string,object>).Add(kv.Key,kv.Value);
}
```

**Reply by jrkumar-conga** [reply]
Thanks for quick reply @abbasc52, I will try this and come back.

**Reply by jrkumar-conga** [reply]
@abbasc52 As expression can be complex, we cannot use the casting and, parsing the expression may introduce latency which we want to avoid as much as possible. the Expando object solution you gave is working fine if the input is Dictionary<string, object> however facing issue when a concrete object is used and if it's having a dictionary<string, object>. We are getting the same` 'Operator '*' incompatible with operand types 'Object' and 'Double'` not sure what is missing, as I have even tried recursively to convert to ExpnadoObject.
e.g. 
```C#
MyObject obj = new {
                           Id = "E678073E-F8CC-E911-B5E9-0003FFE50D88",
                           Name = "WhoAmI",
                           Price = 100,
                           CustomProperties = new Dictionary<string, object>(){
                              { "Id", Guid.NewGuid() },
                              { "CustomPrice", 100 },
                              {"StartDate", DateTime.Now },
                              {"EndDate", DateTime.Now.AddDays(2) }
                }
         }
```
And the expression is like `input1.CustomPrice * 0.5` getting the above mentioned error 
Just to try I have used the code provided by you and written below ToExpando method
```C#
public static ExpandoObject ToExpandoObject<T>(this T obj)
        {
            IDictionary<string, object> expando = new ExpandoObject();
            foreach (var propertyInfo in typeof(T).GetProperties())
            {
                var currentValue = propertyInfo.GetValue(obj);
                if (currentValue is Dictionary<string, object>)
                {
                    dynamic dynamicInputData = new ExpandoObject();
                    foreach (var kv in (Dictionary<string, object>)currentValue)
                    {
                        (dynamicInputData as IDictionary<string, object>).Add(kv.Key, kv.Value);
                    }
                    expando.Add(propertyInfo.Name, dynamicInputData);
                }
                else
                {
                    expando.Add(propertyInfo.Name, currentValue);
                }
            }
            return expando as ExpandoObject;
        }
```

Please let me if we are missing anything or something we are doing wrong.
